File: FactorAnalysis/checkEpsFA.m
function chk = checkEpsFA(lambda,psi,gamma,eps)
%{
This function checks whether the selected epsilon value leads to REM estimates 
that meet the heuristic for hyperparameter selection.
    
INPUT:
    lambda: (p x k) estimated lambda
    psi: (p x 1) estimated psi
    gamma: estimated gamma from REM estimation
    eps: hyperparameter for REM estimation 

OUTPUT:
    chk: probability that gamma.*f <= (1-gamma)*eps 
         or equivalently weights <= 1/2
%}


% Generate some data;
n = 10000;
p = size(lambda,1);
sigma = lambda*lambda' + diag(psi);
V = chol(sigma);
X = randn(n,p)*V;
X = X';

% Calculate likelihood;
logdetV = log(det(V));  % the det(V) is the square root of the det of sigmax;
log_f = -(1/2)*(p*log(2*pi)+2*logdetV+diag((X'/V)/(V')*X));
f = exp(log_f);

% Calculate weight distribution;
weights = gamma*f ./ (gamma*f + (1-gamma)*eps);

% Calculate chk;
chk = mean(weights < 1/2);

    
end

File: FactorAnalysis/R_code/checkEpsFA.R
checkEpsFA <- function(nu, lambda, psi, eps, gamma){
  
  # simulate data
  n = 2e4
  p = nrow(lambda)
  sigma = lambda %*% t(lambda) + diag(psi)
  V = chol(sigma)
  X = sweep(matrix(rnorm(n*p), nrow=n) %*% V, 2, nu, "+")

  # calculate log-likelihoods
  logdetV = log(det(V))
  Z = sweep(X, 2, nu, "-") %*% solve(V)
  ind_lik = -(1/2)*(p*log(2*pi) + 2*logdetV + apply(Z, 1, function(y) t(y) %*% y))
  
  # compute weights
  weights_num = log(gamma) + ind_lik
  weights_denom = log(gamma*exp(ind_lik) + (1-gamma)*eps)
  log_weights = weights_num - weights_denom
  weights = exp(log_weights)
 
  # calculate heuristic check value
  chk = 1 - mean(weights)
  
  return(chk)
}